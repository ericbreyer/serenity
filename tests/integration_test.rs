#![cfg(test)]

use anyhow::Result;
use insta::{assert_debug_snapshot, assert_snapshot};
use test_case::test_case;

#[test_case("trivial.ser", 1; "trivial")]
#[test_case("prime_sieve.ser", 8181807856294299570; "prime_sieve")]
#[test_case("interfaces.ser", 12; "interfaces")]
#[test_case("babbage.ser", 25264; "babbage")]
#[test_case("linkedlist.ser", 55; "linkedlist")]
#[test_case("prime_conspiricy.ser", 7942686168; "prime_conspiricy")]
#[test_case("highly_composites.ser", 355168; "highly_composites")]
pub fn test_file_run(file : &str, ecode: i64) -> Result<()> {

    let file = format!("tests/{}", file);

    let mut out = Vec::new();
    let code = serenity::run_file(&file, &mut out)?;

    assert!(code == ecode, "Expected: {}, Got: {}", ecode, code);    
    
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_scan(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let tokens = serenity::scan(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_debug_snapshot!(tokens);
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_parse(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let ast = serenity::parse(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_snapshot!(ast);
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_compile(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let ir = serenity::compile(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_snapshot!(ir);
    Ok(())
}

