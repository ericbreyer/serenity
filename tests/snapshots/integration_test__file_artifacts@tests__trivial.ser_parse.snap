---
source: tests/integration_test.rs
expression: result
---
[line    1] Function Declaration: 7250923139693181361_next
[line    1] ╰─── Function Expression:
[line    1]      ├─── Name: 7250923139693181361_next
[line    1]      ├─── Return Type: fn[*struct generator_impl { self: *nil, vtable: *struct generator_vtable { next: fn[](*nil) -> fn[*nil]() -> int, }, }]() -> int
[line    1]      ├─── Params:
[line    1]      │    ╰─── self: *struct generator_impl { self: *nil, vtable: *struct generator_vtable { next: fn[](*nil) -> fn[*nil]() -> int, }, }
[line    1]      ╰─── Body: 
[line    1]           ╰─── Return:
[line    0]                ╰─── Function: 7250923139693181361_next_inner
[line    0]                     ╰─── Function Expression:
[line    0]                          ├─── Name: 7250923139693181361_next_inner
[line    0]                          ├─── Return Type: int
[line    0]                          ├─── Captures:
[line    0]                          │    ╰─── self
[line    0]                          ╰─── Body: 
[line    1]                               ╰─── Return:
[line    2]                                    ╰─── Call:
[line    2]                                         ╰─── Call:
[line    2]                                              ├─── Get: "next"
[line    2]                                              │    ╰─── Deref: 
[line    2]                                              │         ╰─── Get: "vtable"
[line    2]                                              │              ╰─── Deref: 
[line    2]                                              │                   ╰─── Variable: <Identifier   "self"        [2]>
[line    2]                                              ╰─── Get: "self"
[line    2]                                                   ╰─── Deref: 
[line    2]                                                        ╰─── Variable: <Identifier   "self"        [2]>
[line    1] Function Declaration: 15292438991726710161_next
[line    1] ╰─── Function Expression:
[line    1]      ├─── Name: 15292438991726710161_next
[line    1]      ├─── Return Type: fn[*struct nat_gen { i: int, }]() -> int
[line    1]      ├─── Params:
[line    1]      │    ╰─── self: *struct nat_gen { i: int, }
[line    1]      ╰─── Body: 
[line    1]           ╰─── Return:
[line    0]                ╰─── Function: 15292438991726710161_next_inner
[line    0]                     ╰─── Function Expression:
[line    0]                          ├─── Name: 15292438991726710161_next_inner
[line    0]                          ├─── Return Type: int
[line    0]                          ├─── Captures:
[line    0]                          │    ╰─── self
[line    1] Var Declaration: nat_gen_generator_vtable
[line    3] Function Declaration: nat_gen_impl_generator
[line    3] ╰─── Function Expression:
[line    3]      ├─── Name: nat_gen_impl_generator
[line    3]      ├─── Return Type: struct generator_impl { self: *nil, vtable: *struct generator_vtable { next: fn[](*nil) -> fn[*nil]() -> int, }, }
[line    3]      ├─── Params:
[line    3]      │    ╰─── self: *struct nat_gen { i: int, }
[line    3]      ╰─── Body: 
[line    4]           ├─── Expression: 
[line    4]           │    ╰─── Assign:
[line    4]           │         ├─── Get: "next"
[line    4]           │         │    ╰─── Variable: <Identifier   "nat_gen_generator_vtable" [4]>
[line    4]           │         ╰─── Variable: <Identifier   "15292438991726710161_next" [4]>
[line    4]           ╰─── Return:
[line    6]                ╰─── StructInitializer: struct generator_impl {self: *nil, vtable: *struct generator_vtable { next: fn[](*nil) -> fn[*nil]() -> int, }, }
[line    6]                     ├───  vtable: 
[line    6]                     │    ╰─── Ref: 
[line    6]                     │         ╰─── Variable: <Identifier   "nat_gen_generator_vtable" [6]>
[line    6]                     ╰───  self: 
[line    6]                          ╰─── Variable: <Identifier   "self"        [6]>
[line    9] Function Declaration: 15292438991726710161_next
[line    9] ╰─── Function Expression:
[line    9]      ├─── Name: 15292438991726710161_next
[line    9]      ├─── Return Type: fn[*struct nat_gen { i: int, }]() -> int
[line    9]      ├─── Params:
[line    9]      │    ╰─── self: *struct nat_gen { i: int, }
[line    9]      ╰─── Body: 
[line    9]           ╰─── Return:
[line    0]                ╰─── Function: 15292438991726710161_next_inner
[line    0]                     ╰─── Function Expression:
[line    0]                          ├─── Name: 15292438991726710161_next_inner
[line    0]                          ├─── Return Type: int
[line    0]                          ├─── Captures:
[line    0]                          │    ╰─── self
[line    0]                          ╰─── Body: 
[line   10]                               ├─── Var Declaration: i
[line   10]                               │    ╰───  Initializer
[line   10]                               │         ╰─── Get: "i"
[line   10]                               │              ╰─── Deref: 
[line   10]                               │                   ╰─── Variable: <Identifier   "self"        [10]>
[line   11]                               ├─── Expression: 
[line   11]                               │    ╰─── Assign:
[line   11]                               │         ├─── Get: "i"
[line   11]                               │         │    ╰─── Deref: 
[line   11]                               │         │         ╰─── Variable: <Identifier   "self"        [11]>
[line   11]                               │         ╰─── Binary: Plus
[line   11]                               │              ├─── Variable: <Identifier   "i"           [11]>
[line   11]                               │              ╰─── Literal: 1
[line   11]                               ╰─── Return:
[line   12]                                    ╰─── Variable: <Identifier   "i"           [12]>
[line   15] Function Declaration: main
[line   15] ╰─── Function Expression:
[line   15]      ├─── Name: main
[line   15]      ├─── Return Type: int
[line   15]      ╰─── Body: 
[line   16]           ├─── Var Declaration: gen
[line   16]           │    ╰───  Initializer
[line   16]           │         ╰─── StructInitializer: struct nat_gen {i: int, }
[line   16]           │              ╰───  i: 
[line   16]           │                   ╰─── Literal: 0
[line   17]           ├─── Var Declaration: g
[line   17]           │    ╰───  Initializer
[line   17]           │         ╰─── Call:
[line   17]           │              ├─── Variable: <Identifier   "nat_gen_impl_generator" [17]>
[line   17]           │              ╰─── Ref: 
[line   17]           │                   ╰─── Variable: <Identifier   "gen"         [17]>
[line   17]           ╰─── Return:
[line   18]                ╰─── Binary: Plus
[line   18]                     ├─── Call:
[line   18]                     │    ╰─── Get: "next"
[line   18]                     │         ╰─── Variable: <Identifier   "g"           [18]>
[line   18]                     ╰─── Call:
[line   18]                          ╰─── Get: "next"
[line   18]                               ╰─── Variable: <Identifier   "g"           [18]>
